java.lang.IllegalArgumentException: Stack should be empty on weirdly positioned @Supress("ktlint")

<!-- The bug you're experiencing might have already been reported!
Please search in the [issues](https://github.com/pinterest/ktlint/issues) before creating one. -->

## Expected Behavior
<!---Tell us what should happen. -->
<!-- If relevant, provide a link to the relevant section in the [Kotlin Coding Conventions](https://kotlinlang.org/docs/coding-conventions.html) or [Android Kotlin Style Guide](https://developer.android.com/kotlin/style-guide).  -->

It should parse the file - it parsed it as of 1.2.1, it doesn't as of 1.7.1.

<!-- Note: Ktlint should not produce any code which conflict with IntelliJ default code formatting. So please verify that your expectation is accepted by IntelliJ default code formatting. -->

## Observed Behavior
<!---Tell us what happens instead of the expected behavior -->
<!--- Provide the exact command which was executed but please -->
<!--- ensure that the flag "--log-level=debug" is added to the -->
<!--- command as well. Provide the output of this command. -->

It breaks with

```
 Caused by: org.gradle.api.GradleException: KtLint failed to parse file: /local/p4clients/pkgbuild-const/workspace/src/AvionInternalLibrary/tst/com/amazon/avion/internal/aws/iam/IAMRoleTest.kt
    at org.jlleitschuh.gradle.ktlint.worker.KtLintWorkAction.execute(KtLintWorkAction.kt:65)
    at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
    at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:54)
    at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:48)
    at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
    at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:48)
    at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
    at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
    at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:108)
    at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:77)
    at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$1(WorkerAction.java:150)
    at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
    at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$2(WorkerAction.java:150)
    at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:85)
    at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:142)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
    at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
    at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
    at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
    at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
    at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
Caused by: java.lang.IllegalArgumentException: Stack should be empty:
    IndentContext(fromASTNode=Element(kotlin.FILE), toASTNode=PsiWhiteSpace, nodeIndent=, firstChildIndent=, childIndent=, lastChildIndent=, activated=true)
    IndentContext(fromASTNode=Element(CLASS), toASTNode=PsiElement(RBRACE), nodeIndent=, firstChildIndent=, childIndent=, lastChildIndent=, activated=false)
    IndentContext(fromASTNode=PsiElement(LBRACE), toASTNode=PsiElement(RBRACE), nodeIndent=, firstChildIndent=, childIndent=    , lastChildIndent=, activated=true)
    IndentContext(fromASTNode=Element(OBJECT_DECLARATION), toASTNode=PsiElement(RBRACE), nodeIndent=    , firstChildIndent=, childIndent=, lastChildIndent=, activated=false)
    IndentContext(fromASTNode=PsiElement(LBRACE), toASTNode=PsiElement(RBRACE), nodeIndent=    , firstChildIndent=, childIndent=    , lastChildIndent=, activated=true)
    IndentContext(fromASTNode=PsiElement(EQ), toASTNode=PsiElement(RPAR), nodeIndent=        , firstChildIndent=    , childIndent=    , lastChildIndent=    , activated=true)
    IndentContext(fromASTNode=PsiElement(EQ), toASTNode=PsiElement(RPAR), nodeIndent=            , firstChildIndent=    , childIndent=    , lastChildIndent=    , activated=true)
    at com.pinterest.ktlint.ruleset.standard.rules.IndentationRule.afterLastNode(IndentationRule.kt:1141)
    at com.pinterest.ktlint.rule.engine.internal.RuleExecutionContext.executeRule(RuleExecutionContext.kt:73)
    at com.pinterest.ktlint.rule.engine.internal.CodeFormatter.executeRule(CodeFormatter.kt:138)
    at com.pinterest.ktlint.rule.engine.internal.CodeFormatter.format(CodeFormatter.kt:109)
    at com.pinterest.ktlint.rule.engine.internal.CodeFormatter.format(CodeFormatter.kt:58)
    at com.pinterest.ktlint.rule.engine.internal.CodeFormatter.format(CodeFormatter.kt:28)
    at com.pinterest.ktlint.rule.engine.api.KtLintRuleEngine.lint(KtLintRuleEngine.kt:91)
    at org.jlleitschuh.gradle.ktlint.worker.KtLintInvocation100.invokeLint(KtLintInvocation100.kt:170)
    at org.jlleitschuh.gradle.ktlint.worker.KtLintWorkAction.execute(KtLintWorkAction.kt:60)
    ... 24 more

```

This feels similar to #2695. Looking at the IndentationRule, I see this was added here - https://github.com/pinterest/ktlint/commit/b193e5b53ac653c428d56a53fed9d0c971cb7a6d#diff-0153b6178f7ac64e3465ecb22a2c129cc911b3ad28fc5b2185faa2f5cbf28d15R655 - as part of #2119. 

AI comment on the issue:

> The activated = true parameter is commented out on line 1141. This means that when there's a primary constructor with the constructor keyword in ktlint_official code style, the indent context is created but not activated, which could cause incorrect indentation behavior.
The bug is that the indentation context for the primary constructor should be activated immediately when using ktlint_official style, but the activated = true parameter is commented
out, so it defaults to false. This would cause the indentation to not be applied correctly until the next newline activates it.

I agree that this looks like really ugly code, but fact is that it did compile on 1.2.1 and it doesn't in 1.7.1 and I'm not sure why tbh.

## Steps to Reproduce
<!--- Provide a code example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Minimize the example to the bare -->
<!--- minimum required to reproduce the bug. Small, well written -->
<!--- example are proven to be resolved quicker. -->

```
class IAMRoleTest {
    companion object {
        @JvmStatic
        fun prepareToARNAWithStageArguments() = @Suppress("ktlint")
        listOf(
            Arguments.of(IAMRole.CONFIG_GENERATOR_TASK,           Stage.Alpha,   buildARNString(AWSAccount.ALPHA.id, "ConfigGeneratorTaskRole")),
        )
    }
}
```

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Version of ktlint used:
* Relevant parts of the `.editorconfig` settings
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task):
* Version of Gradle used (if applicable):
* Operating System and version:

--- a/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/IndentationRule.kt
+++ b/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/rules/IndentationRule.kt
@@ -1115,7 +1115,14 @@ public class IndentationRule :
         node: ASTNode,
         emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> AutocorrectDecision,
     ) {
-        while (indentContextStack.peekLast()?.toASTNode == node) {
+        while (
+            // End visit of the node matches with the toAstNode on top of the stack
+            indentContextStack.peekLast()?.toASTNode == node ||
+            // When the ktlint indentation violations on child elements have been suppressed, those child elements are not visited, and the
+            // stack is therefore not cleaned up properly. So remove all elements from top of the stack until the start of an index context
+            // is found for the current node.
+            indentContextStack.peekLast()?.toASTNode?.parent { it == node } != null
+        ) {
             LOGGER.trace {
                 val indentContext = indentContextStack.peekLast()
                 val nodeIndentLevel = indentConfig.indentLevelFrom(indentContext.nodeIndent)

--- a/ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/IndentationRuleTest.kt
+++ b/ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/rules/IndentationRuleTest.kt
@@ -5932,6 +5932,24 @@ internal class IndentationRuleTest {
         }
     }
 
+    @ParameterizedTest(name = "Suppression: {0}")
+    @ValueSource(
+        strings = [
+            "ktlint",
+            "ktlint:standard:indent",
+        ],
+    )
+    fun `Issue 3109 - Suppressing all ktlint violations may not result in failing to empty the stack`(suppression: String) {
+        val code =
+            """
+            fun foo() =
+                @Suppress("${suppression}")
+                listOf("foo")
+            """.trimIndent()
+
+        indentationRuleAssertThat(code).hasNoLintViolations()
+    }
+
     private companion object {
         val INDENT_STYLE_TAB =
             INDENT_STYLE_PROPERTY to PropertyType.IndentStyleValue.tab

